### PR TITLE
Introduce callback via Pybind11 to allow logging

### PR DESCRIPTION
**Pull request by BjarneHerland**

### CODE REVIEW

**Comment by BjarneHerland** [comment]
PoC for how we can do Python-logging from C++. See rationale and description in https://github.com/equinor/ert/issues/2353

**Comment by BjarneHerland** [comment]
Mmm... heavy changes on main... 😬    Rebasing on Monday...

**Comment by BjarneHerland** [comment]
test ert please

**Comment by oyvindeide** [comment]
This looks like an elegant solution to me, but maybe @dotfloat could also have a look?

**Comment by dafeda** [comment]
I think this looks great, but am not comfortable approving it due to weak C++.
@dotfloat and @eivindjahren 🙏

**Comment by BjarneHerland** [comment]
Yes, @dotfloat and @eivindjahren please have a look! I see two possible weaknesses:

1. How does it work if we spin out a new process from C++ and/or Python after setting the log-callback? In Python I expect we have some solution for logging in general which we could extend with this. In C++ I really don't know.
2. Ref `main.py` in the patch: Why is it necessary to set the handler to None in the end? The program hangs otherwise, and is what makes the `if callback` statements necessary in `cpp_logger.hpp` (for tests to pass). When thinking about it, this could proabably be achieved by wrapping the static callback instead and fall back to NOOP-lambda...

**Comment by BjarneHerland** [comment]
> 2\. is what makes the `if callback` statements necessary in `cpp_logger.hpp` (for tests to pass). When thinking about it, this could proabably be achieved by wrapping the static callback instead and fall back to NOOP-lambda...

Fixed this in newest update

**Comment by eivindjahren** [comment]
> 1. How does it work if we spin out a new process from C++ and/or Python after setting the log-callback? In Python I expect we have some solution for logging in general which we could extend with this. In C++ I really don't know.

I defer to @dotfloat on that issue as I am quite unsure about this myself.

> 2. Ref `main.py` in the patch: Why is it necessary to set the handler to None in the end? The program hangs otherwise, and is what makes the `if callback` statements necessary in `cpp_logger.hpp` (for tests to pass). When thinking about it, this could proabably be achieved by wrapping the static callback instead and fall back to NOOP-lambda...

Again, not sure, but it seems to be some resource management issue. We might be chasing a red harring by looking at the cpp logger as it might a deadlock caused by the ert server.

**Comment by BjarneHerland** [comment]
> Again, not sure, but it seems to be some resource management issue. We might be chasing a red harring by looking at the cpp logger as it might a deadlock caused by the ert server.

Simple tests also hangs unless we set the callback to `None`. I suspect some interaction between Pybind11 and GC in Python but have not spent time looking into it, hoping that one of you guys know this already 😄

**Comment by eivindjahren** [comment]
> Simple tests also hangs unless we set the callback to `None`. I suspect some interaction between Pybind11 and GC in Python but have not spent time looking into it, hoping that one of you guys know this already smile

Perhaps by [`gc.set_debug(gc.DEBUG_INSTANCES)`](https://docs.python.org/3/library/gc.html#gc.set_debug) you can have a look at why the logger is not released.

**Comment by eivindjahren** [comment]
This seems relevant: https://github.com/pybind/pybind11/issues/1866#issuecomment-526948586

**Comment by BjarneHerland** [comment]
> This seems relevant: [pybind/pybind11#1866 (comment)](https://github.com/pybind/pybind11/issues/1866#issuecomment-526948586)

Absolutely! Thanks - updated code...

**Comment by eivindjahren** [comment]
@dotfloat should this be connected to the `util_abort` handler?

**Comment by dotfloat** [comment]
I can't say I'm a big fan of this. I would use [spdlog](https://github.com/gabime/spdlog) and add a spdlog handler that redirects it to Python's logging. There's no need to do this callback stuff.

**Comment by eivindjahren** [comment]
> 

Interesting, is spdlog available through conan?

**Comment by dotfloat** [comment]
`spdlog` is basically _the_ modern C++ logging library. So yes.

File: tests/libres_tests/res/enkf/test_enkf.py [hunk 1]
```diff
@@ -74,6 +77,30 @@ def test_bootstrap(self):
             main = EnKFMain(res_config)
             self.assertTrue(main, "Load failed")
 
+    @tmpdir()
```
Comment (dotfloat) [inline tests/libres_tests/res/enkf/test_enkf.py:80]: Don't create new old-style tests. Create a new pytest test at the bottom of the file.

**Comment by BjarneHerland** [comment]
> I can't say I'm a big fan of this. I would use [spdlog](https://github.com/gabime/spdlog) and add a spdlog handler that redirects it to Python's logging. There's no need to do this callback stuff.

I did consider 3rd+paty C++ loggers, including spdlog. My conclusion was that 1) we are only interested in forwarded log-entries, logging from C++ as such is not useful to us, 2) it's not obvious (to me) how to install a handler without initiating this from the Python-environment, which is what we do now anyway, 3) we avoid an extra dependency by this thin layer.

Having said this, I am as always open to suggestions! 😄   In particular to point 2 above.

File: tests/libres_tests/res/enkf/test_enkf.py (hunk 1 above)
Comment (BjarneHerland) [reply tests/libres_tests/res/enkf/test_enkf.py:80]: Perhaps a new testfile in `libres_tests/res/config` ?  What would be a good template for a modern test?

**Comment by dotfloat** [comment]
> I did consider 3rd+paty C++ loggers, including spdlog. My conclusion was that 1) we are only interested in forwarded log-entries, logging from C++ as such is not useful to us, 2) it's not obvious (to me) how to install a handler without initiating this from the Python-environment, which is what we do now anyway, 3) we avoid an extra dependency by this thin layer.
> 
> Having said this, I am as always open to suggestions! smile In particular to point 2 above.

It seems that in this PR you're logging something in the C implementation of res config which you expect to see in Python, so I'm not entirely sure what you mean by 1. For 3, I think it's fine to have an extra dependency, especially if it's tiny, so that we don't fall into every C programmer's trap of reinventing the universe.

Here's my attempt: https://github.com/equinor/ert/pull/2550

**Comment by BjarneHerland** [comment]
> Here's my attempt: #2550

Very nice! I After my comment #2 it struck me also that is should be possible to expose spdlog via Pybind11, but the trick of grabbing the logger in `init_spdlog()` is nice.

My meaning with 1 was that we don't need a full-featured log-module from C++, just a simple forwarder.

I suggest to close this one and I'll review your attempt.

File: tests/libres_tests/res/enkf/test_enkf.py (hunk 1 above)
Comment (eivindjahren) [reply tests/libres_tests/res/enkf/test_enkf.py:80]: Have a look at `ert_tests/ert3` for some good new style tests.

Comment (eivindjahren) [reply tests/libres_tests/res/enkf/test_enkf.py:80]: As you can see here: https://github.com/equinor/ecl/blob/3d6d17e1470419fc5af0c891560fb2211b2eda0d/python/tests/ecl_tests/test_ecl_sum.py
A file can both contain tests in a class and function tests (both the members of `EclSumTest` and `test_write_repr` will be run).

**Comment by BjarneHerland** [comment]
> I suggest to close this one and I'll review your attempt.

Refer to https://github.com/equinor/ert/pull/2550#issuecomment-993311408

**Pull request closed without merging**